src/ProtobufServer/ecdar_requests/send_query.rs

```diff
@@ -142,327 +143,58 @@ fn create_components(components: Vec<Component>) -> HashMap<String, Component> {
-                component_name: locations1.id.get_component_id().unwrap(),
-                component_index: 0,
-            }),
-        },
-        Location {
-            id: locations2.id.to_string(),
-            specific_component: Some(SpecificComponent {
-                component_name: locations2.id.get_component_id().unwrap(),
-                component_index: 0,
-            }),
-        },
-    ];
-    loc_vec
+impl IntoProtoResult for SystemRecipeFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Model(self.into())
+    }
 }
 
-fn make_proto_zone(state_pair: &StatePair) -> Option<Federation> {
-    let disjunction = state_pair.ref_zone().minimal_constraints();
-    let mut conjunctions: Vec<ProtobufConjunction> = vec![];
-    for conjunction in disjunction.conjunctions.iter() {
-        let mut constraints: Vec<ProtobufConstraint> = vec![];
-        for constraint in conjunction.constraints.iter() {
-            constraints.push(ProtobufConstraint {
-                x: Some(ProtobufComponentClock {
-                    specific_component: value_in_component(Some(&state_pair.locations1.id)),
-                    clock_name: constraint.i.to_string(),
-                }),
-                y: Some(ProtobufComponentClock {
-                    specific_component: value_in_component(Some(&state_pair.locations1.id)),
-                    clock_name: constraint.j.to_string(),
-                }),
-                strict: constraint.ineq().is_strict(),
-                c: constraint.ineq().bound(),
-            });
-        }
-        conjunctions.push(ProtobufConjunction { constraints })
+impl IntoProtoResult for DeterminismFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Determinism(self.into())
     }
-    Some(Federation {
-        disjunction: Some(ProtobufDisjunction { conjunctions }),
-    })
 }
 
-fn value_in_location(maybe_location: &Option<LocationID>) -> String {
-    match maybe_location {
-        Some(location_id) => location_id.to_string(),
-        None => "".to_string(),
+impl IntoProtoResult for ConsistencyFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Consistency(self.into())
     }
 }
 
-fn value_in_component(maybe_location: Option<&LocationID>) -> Option<SpecificComponent> {
-    maybe_location.map(|location_id| SpecificComponent {
-        component_name: location_id.get_component_id().unwrap(),
-        component_index: 0,
-    })
+impl IntoProtoResult for RefinementFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Refinement(self.into())
+    }
 }
 
-fn value_in_action(maybe_action: &Option<String>) -> String {
-    match maybe_action {
-        Some(action) => action.to_string(),
-        None => "".to_string(),
+impl IntoProtoResult for PathFailure {
+    fn into_proto_result(self) -> ProtobufResult {
+        ProtobufResult::Reachability(self.into())
     }
 }
```

Comment:
I think that there is a lot of duplicate(ish) code here. This could be removed by using a macro instead:
```suggestion
macro_rules! impl_intoProtoResult {
    ($T:ident, $variant:ident) => {
        impl IntoProtoResult for $T {
            fn into_proto_result(self) -> ProtobufResult {
                crate::ProtobufServer::services::query_response::Result::$variant(self.into())
            }
        }
    };
}
impl_intoProtoResult!(SystemRecipeFailure, Model);
impl_intoProtoResult!(DeterminismFailure, Determinism);
impl_intoProtoResult!(ConsistencyFailure, Consistency);
impl_intoProtoResult!(RefinementFailure, Refinement);
impl_intoProtoResult!(PathFailure, Reachability);

```
While this is not a very pretty solution, I still think it is more concise. I'll let you decide :smile:

Reply:
I can see how the macro is more concise but i do think the macro is much less readable. I sort of don't think the code is "ugly"/long enough to warrant a macro with multiple arguments.

We could consider moving the trait implementations to a separate file?

Sort of unrelated as the duplication remains; I have now switched to implementing `From`/`Into` instead of defining the new `IntoProtoResult` trait which has the exact same functionality as `impl Into<ProtoResult>`.

Reply:
I sort of like the idea of having them in a separate file, though I feel like it is hard to justify it, so I think we should just leave it here.